### PR TITLE
Reimplement and improve CLI

### LINKED ISSUE #53: Clean up CLI parameter parsing
Robyn suggested this as a cleaner way to have parameters and help messages

### LINKED ISSUE #226: Incorrect VCF Number used for some FORMAT fields
According to section "1.4 Meta-information lines" from the VCF 4.3 [specs](norm.sorted.bcf) `Number=A` means

> A: The field has one value per alternate allele.  The values must be in the same order as listed in the ALT column.

Take this example header line and variant

```
##FORMAT=<ID=MEAN_FWD_COVG,Number=A,Type=Integer,Description="Mean forward coverage">
chrom    1350    .       AA      GA      .       .       SVTYPE=PH_SNPs;GRAPHTYPE=SIMPLE GT:MEAN_FWD_COVG:MEAN_REV_COVG:MED_FWD_COVG:MED_REV_COVG:SUM_FWD_COVG:SUM_REV_COVG:GAPS:LIKELIHOOD:GT_CONF      .:16,16:5,5:16,16:5,5:16,16:5,5:0,0:-104.622,-104.622:0
```

We actually use `MEAN_FWD_COVG` (and many other FORMAT fields) as if it is Number R

> R: The field has one value for each possible allele, including the reference.  The order of the values must be the reference allele first, then the alternate alleles as listed in the ALT column.

So, when trying to run a pandora VCF through `bcftools view -a <in.vcf>` I receive the following error message

```
[E::bcf_remove_allele_set] Unexpected number of values in FORMAT/MEAN_FWD_COVG at chrom:1350; expected Number=A=1, but found 2
Error: Could not trim alleles at chrom:1350
```

### PR DESCRIPTION
**Pull request by mbhall88**
This PR moves the handling of the Command-line interface (CLI) to [CLI11] and closes #53.

Additionally, some changes in this PR are not backwards compatible as some options have been renamed/removed.

This PR also contains a hotfix that closes #226 

[CLI11]: https://github.com/CLIUtils/CLI11

### CODE REVIEW

**Comment by mbhall88** [comment]
# Index

# Old

```
$ pandora index -h
Usage: pandora index [options] <prgs.fa>
Options:
        -h,--help                       Show this help message
        -w W                            Window size for (w,k)-minimizers, default 14
        -k K                            K-mer size for (w,k)-minimizers, default 15
        -t T                            Number of threads, default 1
        --offset                                Offset for PRG ids, default 0
        --outfile                               Filename for index
        --log_level                     debug,[info],warning,error
```

# New

```
$ pandora index -h
Index population reference graph (PRG) sequences.
Usage: pandora index [OPTIONS] <PRG>

Positionals:
  <PRG> FILE [required]       PRG to index (in fasta format)

Options:
  -h,--help                   Print this help message and exit
  -w INT                      Window size for (w,k)-minimizers (must be <=k) [default: 14]
  -k INT                      K-mer size for (w,k)-minimizers [default: 15]
  -t,--threads INT            Maximum number of threads to use [default: 1]
  --offset INT                Offset for PRG ids [default: 0]
  -o,--outfile FILE           Filename for the index [default: <PRG>.k15.w14.idx]
  -v                          Verbosity of logging. Repeat for increased verbosity
```

The only point of discussion here I think (which is the same across all subcommands) is the change to the log level indication. Instead of asking the user to state the level explicitly (i.e. `info`, `debug` etc.) it is assumed as `info` and if they want more logging they pass `-v` to indicate `debug` and `-vv` to indicate `trace`.  
Note: this change also addresses and closes #160 

Another nice thing about using CLI11 is there are some nice built-in validators. For this subcomand, if`<PRG>` does not exist, CLI11 throws an exception for us.

**Comment by mbhall88** [comment]
# Map

## Old

```
$ pandora map -h
only provided 2arguments
Usage: pandora map -p PRG_FILE -r READ_FILE <option(s)>
Options:
        -h,--help                       Show this help message
        -p,--prg_file PRG_FILE          Specify a fasta-style prg file
        -r,--read_file READ_FILE        Specify a file of reads in fasta format
        -o,--outdir OUTDIR      Specify directory of output
        -w W                            Window size for (w,k)-minimizers, must be <=k, default 14
        -k K                            K-mer size for (w,k)-minimizers, default 15
        -m,--max_diff INT               Maximum distance between consecutive hits within a cluster, default 500 (bps)
        -e,--error_rate FLOAT           Estimated error rate for reads, default 0.11
        -t T                            Number of threads, default 1
        --genome_size   NUM_BP  Estimated length of genome, used for coverage estimation
        --output_kg                     Save kmer graphs with fwd and rev coverage annotations for found localPRGs
        --output_vcf                    Save a vcf file for each found localPRG
        --vcf_refs REF_FASTA            A fasta file with an entry for each LocalPRG giving reference sequence for
                                        VCF. Must have a perfect match in the graph and the same name as the graph
        --output_comparison_paths       Save a fasta file for a random selection of paths through localPRG
        --output_covgs  Save a file of covgs for each localPRG present, one number per base of fasta file
        --output_mapped_read_fa Save a file for each gene containing read parts which overlapped it
        --illumina                      Data is from illumina rather than nanopore, so is shorter with low error rate
        --clean                 Add a step to clean and detangle the pangraph
        --bin                   Use binomial model for kmer coverages, default is negative binomial
        --max_covg                      Maximum average coverage from reads to accept
        --genotype MODE                 Add extra step to carefully genotype sites. Has two modes: global (ML path oriented) or local (coverage oriented)
        --snps_only                     When genotyping, include only snp sites
        -d,--discover                   Add denovo discovery
        --denovo_kmer_size                      Kmer size to use for denovo discovery
        --log_level                     debug,[info],warning,error
```

## New

```
$ pandora map -h
Quasi-map reads from a single sample to an indexed PRG and infer the presence or absence of PRG loci in the sample. Infers the sequence at loci as a mosaic of reference sequences an
d outputs a set of variant calls for the sample with respect to the graph and optionally identifies de novo variants not present in the graph.
Usage: ./pandora map [OPTIONS] <TARGET> <QUERY>

Positionals:
  <TARGET> FILE [required]    An indexed PRG file (in fasta format)
  <QUERY> FILE [required]     Fast{a,q} file containing reads to quasi-map

Options:
  -h,--help                   Print this help message and exit
  -w INT                      Window size for (w,k)-minimizers (must be <=k) [default: 14]
  -k INT                      K-mer size for (w,k)-minimizers [default: 15]
  -o,--outdir DIR             Directory to write output files to [default: pandora]
  -t,--threads INT            Maximum number of threads to use [default: 1]
  --offset INT                Offset for PRG ids [default: 0]
  --vcf-refs FILE             Fasta file with an entry for each loci. The entries will be used as the reference sequence for their respective loci. The sequence MUST have a perfect
match in <TARGET> and the same name
  -e,--error-rate FLOAT       Estimated error rate for reads [default: 0.11]
  -g,--genome-size INT        Estimated length of the genome - used for coverage estimation [default: 5000000]
  -m,--max-diff INT           Maximum distance (bp) between consecutive hits within a cluster [default: 250]
  --output-kg                 Save kmer graphs with forward and reverse coverage annotations for found loci
  --output-vcf                Save a VCF file for each found loci
  --output-comparison-paths   Save a fasta file for a random selection of paths through loci
  --output-covgs              Save a file of coverages for each loci present - one number per base
  --output-mapped-read-fa     Save a file for each loci containing read parts which overlapped it
  --illumina                  Reads are from Illumina. Alters error rate used and adjusts for shorter reads
  --clean                     Add a step to clean and detangle the pangraph
  --bin                       Use binomial model for kmer coverages [default: negative binomial]
  --max-covg INT              Maximum average coverage of reads to accept [default: 300]
  --genotype MODE             Add extra step to carefully genotype sites. There are two modes: 'global' (ML path oriented) or 'local' (coverage oriented)
  --snps                      When genotyping, only include SNP sites
  -v                          Verbosity of logging. Repeat for increased verbosity
  -d,--discover               Add a step to discover de novo variants
  --discover-k INT            K-mer size to use when disovering de novo variants [default: 11]
  -c,--min-cluster-size INT   Minimum size of a cluster of hits between a read and a loci to consider the loci present [default: 10]
  --kmer-avg INT              Maximum number of kmers to average over when selecting the maximum likelihood path [default: 100]
  --min-kmer-covg INT         Should this be exposed?
  --min-allele-covg-gt INT    Should this be exposed?
  --min-total-covg-gt INT     Should this be exposed?
  --min-diff-covg-gt INT      Should this be exposed?
  --min-allele-fraction-covg-gt INT
                              Should this be exposed?
  --genotyping-error-rate FLOAT
  --confidence-threshold INT
```

There is a lot to address here. In particular, I have added "Should this be exposed?" to some options. If we do want these options exposed, then a description is needed for them. Additionally, I would like to shorten a lot of these options as some are far too verbose. I have left two options without a description - `--confidence-threshold` and `--genotyping-error-rate` as I struggled to determine what these do from the code.

One outstanding thing is to create groups within the menu (similar to tools like `minimap2` and `bwa`) so we would have something like "Input/Output Options", "Mapping", "Preset" etc. Please feel free to suggest groups. I will wait until we have settled on what options should be removed (if any).

A breaking change here also is removing the snakecase convention in options in favour of the more widely used "kebab case". i.e. `--confidence_threshold` becomes `--confidence-threshold`. If this is a big issue I will change it back, but I *really* dislike snake case in CLI opts.

Lastly, I have added 'todo' comments above any options that I have questions or concerns about (`map_main.cpp` function `setup_map_subcommand`) so could we please discuss those in the code review? In particular, there is a comment on line 123 relating to a discussion @iqbal-lab and I had about whether we just move purely to global genotyping.

**Comment by mbhall88** [comment]
# Main

## Old

```
$ pandora --help
[main] unrecognized command '--help'
```

## New

```
$ pandora --help
Pandora, Pan-genome inference and genotyping with long noisy or short accurate reads.
Usage: ./pandora [OPTIONS] SUBCOMMAND

Options:
  -h,--help                   Print this help message and exit

Subcommands:
  index                       Index population reference graph (PRG) sequences.
  map                         Quasi-map reads from a single sample to an indexed PRG and infer the presence or absence of PRG loci in the sample. Infers the sequence at loci as a mosaic of reference sequences and outputs a set of variant calls for the sample with respect to the graph and optionally identifies de novo variants not present in the graph.
```

Note: I have only implemented two subcommands so far, but they will all be visible here when all is done, and I will update this comment as I add more.   
Closes #159

**Comment by mbhall88** [comment]
# Compare

## Old

```
$ pandora compare -h
Usage: pandora compare -p PRG_FILE -r READ_INDEX -o OUTDIR <option(s)>
Options:
        -h,--help                       Show this help message
        -p,--prg_file PRG_FILE          Specify a fasta-style prg file
        -r,--read_index READ_INDEX      Specify a file with a line per sample
                                        sample_id <tab> filepath to reads in fasta/q format
        -o,--outdir OUTDIR      Specify directory of output
        -w W                            Window size for (w,k)-minimizers, default 14
        -k K                            K-mer size for (w,k)-minimizers, default 15
        -m,--max_diff INT               Maximum distance between consecutive hits within a cluster, default 250 (bps)
        -e,--error_rate FLOAT           Estimated error rate for reads, default 0.11
        -t T                            Number of threads, default 1
        --genome_size   NUM_BP  Estimated length of genome, used for coverage estimation
        --vcf_refs REF_FASTA            A fasta file with an entry for each LocalPRG giving reference sequence for
                                        VCF. Must have a perfect match in the graph and the same name as the graph
        --illumina                      Data is from illumina rather than nanopore, so is shorter with low error rate
        --clean                 Add a step to clean and detangle the pangraph
        --bin                   Use binomial model for kmer coverages, default is negative binomial
        --max_covg                      Maximum average coverage from reads to accept
        --genotype MODE                 Add extra step to carefully genotype sites. Has two modes: global (ML path oriented) or local (coverage oriented)
        --log_level                     debug,[info],warning,error
```

## New

```
$ pandora compare -h
Quasi-map reads from multiple samples to an indexed PRG, infer the sequence of present loci in each sample, and call variants between the samples.
Usage: ./pandora compare [OPTIONS] <TARGET> <QUERY_IDX>

Positionals:
  <TARGET> FILE [required]    An indexed PRG file (in fasta format)
  <QUERY_IDX> FILE [required] A tab-delimited file where each line is a sample identifier followed by the path to the fast{a,q} of reads for that sample

Options:
  -h,--help                   Print this help message and exit
  -w INT                      Window size for (w,k)-minimizers (must be <=k) [default: 14]
  -k INT                      K-mer size for (w,k)-minimizers [default: 15]
  -o,--outdir DIR             Directory to write output files to [default: pandora]
  -t,--threads INT            Maximum number of threads to use [default: 1]
  --vcf-refs FILE             Fasta file with an entry for each loci. The entries will be used as the reference sequence for their respective loci. The sequence MUST have a perfect match in <TARGET> and the same name
  -e,--error-rate FLOAT       Estimated error rate for reads [default: 0.11]
  -g,--genome-size INT        Estimated length of the genome - used for coverage estimation [default: 5000000]
  -m,--max-diff INT           Maximum distance (bp) between consecutive hits within a cluster [default: 250]
  --output-vcf                Save a VCF file for each found loci
  --illumina                  Reads are from Illumina. Alters error rate used and adjusts for shorter reads
  --clean                     Add a step to clean and detangle the pangraph
  --bin                       Use binomial model for kmer coverages [default: negative binomial]
  --max-covg INT              Maximum average coverage of reads to accept [default: 300]
  --genotype MODE             Add extra step to carefully genotype sites. There are two modes: 'global' (ML path oriented) or 'local' (coverage oriented)
  -c,--min-cluster-size INT   Minimum size of a cluster of hits between a read and a loci to consider the loci present [default: 10]
  --kmer-avg INT              Maximum number of kmers to average over when selecting the maximum likelihood path [default: 100]
  --min-allele-covg-gt INT    Should this be exposed?
  --min-total-covg-gt INT     Should this be exposed?
  --min-diff-covg-gt INT      Should this be exposed?
  --min-allele-fraction-covg-gt INT
                              Should this be exposed?
  --genotyping-error-rate FLOAT
  --confidence-threshold INT
  -v                          Verbosity of logging. Repeat for increased verbosity
```

The changes here are similar to `map`. However, I noticed that whilst there are a lot of options in common, some aren't exposed in `compare` but *are* exposed in `map` (e.g. `--min-kmer-covg`)

**Comment by leoisl** [comment]
Thanks a lot for this, looks really nice! Thanks for taking the time to improve the CLI!

I have some comments below, but I realise that I am asking some stuff that might take considerable development time, so feel free to ignore everything. Also what you are doing already improve the CLI by a lot, so I think it is already enough. What I am proposing would be towards cleaning more the CLI and eventually breaking commands into more subcommands... but I don't think working a lot on CLI is a high priority task in our TODO list. I would be happy to review and merge your changes, and consider that CLI work is done.

Regarding `pandora index`:
>`   --offset INT                Offset for PRG ids [default: 0]`

I feel like this parameter was sth required so that we could split the PRG into several subfiles, and launch a `pandora index` job for each subfile, and then combine the index later. I guess this is not required anymore, as `pandora index` takes a couple of hours? There are still also performance improvements we can do in `pandora index` if we think it is not quick enough.

Regarding `pandora map/compare`:

>-w INT                      Window size for (w,k)-minimizers (must be <=k) [default: 14]
>-k INT                      K-mer size for (w,k)-minimizers [default: 15]

This is actually not required, given that the `w` and `k` used to build the index is in the index filename. I actually think it is error prone to ask for `w` and `k` in `pandora map/compare`. User might set a specific value for `w` and `k` when running `pandora index` and do not put this value for `pandora map/compare`. Anyway, it is an information that is better obtained by parsing the index filename instead of asking for it.
PS: I am not sure this belongs to this issue. It might be better to create an issue for this;
PS2: this is not major, so not high priority. I actually would prefer to have a metadata file along with the index, instead of parsing metadata from filenames.

Regarding subcommands breakdown:

 I think pandora should produce unfiltered VCFs. Then we should have a `pandora view` or `pandora filter` command to filter VCFs accordingly. Then some parameters might shift into this command... But I am not sure how many parameters we can move, as many of them affects the ML path, so it might not be possible...

There are some more subcommands breakdown, but I don't think will be productive to discuss them now...

**Comment by mbhall88** [comment]
# Walk

## Old

```
$ pandora walk -h
Usage: pandora walk <in_prg.fa> [<seq.fa> | --top | --bottom]
```

## New

```
$ pandora walk -h
Outputs a path through the nodes in a PRG corresponding to the either an input sequence (if it exists) or the top/bottom path
Usage: ./pandora walk [OPTIONS] <PRG>

Positionals:
  <PRG> FILE:Output a path  [required]
                              A PRG file (in fasta format)

Options:
  -h,--help                   Print this help message and exit
  -i,--input FILE Excludes: --top --bottom
                              Fast{a,q} of sequences to output paths through the PRG for
  -T,--top Excludes: --input --bottom
                              Output the top path through each local PRG
  -B,--bottom Excludes: --input --top
                              Output the bottom path through each local PRG
```

**Comment by iqbal-lab** [comment]
re this
>This is actually not required, given that the w and k used to build the index is in the index filename. I actually think it is error prone to ask for w and k in pandora map/compare. User might set a specific value for w and k when running pandora index and do not put this value for pandora map/compare. Anyway, it is an information that is better obtained by parsing the index filename instead of asking for it.

Depending on the filename is better , but also error prone  as people fiddle with filenames - inside a header would be safest. happy to postpone this though

**Comment by mbhall88** [comment]
Re: removing `--offset` from `index`. I will do this as it is a simple change.

Re: removing `k` and `w` I am a little torn about this. I don't like the idea of getting it from the filename, as the user has the option to name the index something else with `--outfile`. As Zam mentions, we could somehow encode it in the index header? Although, as an example, `minimap2` leaves it to the user to ensure they use the same present, `-x`, for both indexing and mapping. So it wouldn't be unusual to maintain the current functionality.

Speaking of the `index` `--outfile` parameter, I think this option is actually useless? Regardless of what it is named I think `pandora` always looks for an expected name? There are two `load` functions, but [this](https://github.com/mbhall88/pandora/blob/e497912ab645f58daf1da84cfdab6bed2983b591/src/index.cpp#L85-L86) is always called.

**Comment by mbhall88** [comment]
## Todo

- [ ] Remove `--offset` from `index`
- [ ] Add [groups](https://github.com/CLIUtils/CLI11/blob/master/examples/groups.cpp) to `map` and `compare`
- [ ] Add subcommand `check_kmergraph`
- [ ] Add subcommand `get_vcf_ref`
- [ ] Add subcommand `random_path`
- [ ] Add subcommand `merge_index`

**Comment by leoisl** [comment]
> Speaking of the `index` `--outfile` parameter, I think this option is actually useless? Regardless of what it is named I think `pandora` always looks for an expected name? There are two `load` functions, but [this](https://github.com/mbhall88/pandora/blob/e497912ab645f58daf1da84cfdab6bed2983b591/src/index.cpp#L85-L86) is always called.

Yeah, I think `--outfile` is not very useful. I think https://github.com/mbhall88/pandora/blob/e497912ab645f58daf1da84cfdab6bed2983b591/include/index.h#L37 is a helper function, and should be made private, whereas https://github.com/mbhall88/pandora/blob/e497912ab645f58daf1da84cfdab6bed2983b591/include/index.h#L35 is in the public interface.

There is a refactoring of the code that I have been postponing since the beginning, which is applying encapsulation to the code (i.e. hiding all attributes, providing access to them through getters/setters, and hide helper methods, just exposing the minimal public interface to classes clients). It will make things a lot easier to change and maintain later (i.e. many parts of the code access attributes directly, and thus when changing, for e.g. a `std::map` to an `IntervalTree` data structure as we did at some point to speed things up, we actually had to change a lot of the code because of this direct access. Encapsulation would make this a lot easier).
I am still unsure about doing this refactoring. I think refactoring code is great and should be a constant effort, but there absolutely no research progress, it is basically code maintenance, and refactoring can be quite lengthy and require quite a big effort. This might be better discussed in a meeting after the paper on how much time to devote for code maintenance, and adding features to pandora.

**Comment by mbhall88** [comment]
> There is a refactoring of the code that I have been postponing since the beginning, which is applying encapsulation to the code (i.e. hiding all attributes, providing access to them through getters/setters, and hide helper methods, just exposing the minimal public interface to classes clients). It will make things a lot easier to change and maintain later (i.e. many parts of the code access attributes directly, and thus when changing, for e.g. a `std::map` to an `IntervalTree` data structure as we did at some point to speed things up, we actually had to change a lot of the code because of this direct access. Encapsulation would make this a lot easier).
> I am still unsure about doing this refactoring. I think refactoring code is great and should be a constant effort, but there absolutely no research progress, it is basically code maintenance, and refactoring can be quite lengthy and require quite a big effort. This might be better discussed in a meeting after the paper on how much time to devote for code maintenance, and adding features to pandora.

https://www.artima.com/intv/fixit.html

**Comment by mbhall88** [comment]
Group names are added to https://github.com/rmcolq/pandora/pull/224#issuecomment-684176123. Let me know if anything needs changing

**Comment by mbhall88** [comment]
# check_kmergraph

# Old

```
Usage: pandora check_kmergraph <prg.fa> <seq.fa> <k> <w> <flag>
```

# New

```
$ pandora check_kmergraph -h
For each sequence, return the path through the PRG
Usage: pandora check_kmergraph [OPTIONS] <PRG>

Positionals:
  <PRG> FILE [required]       PRG to index (in fasta format)

Options:
  -h,--help                   Print this help message and exit
  -i,--input FILE Excludes: --top --bottom
                              Fast{a,q} of sequences to output paths through the PRG for
  -w INT                      Window size for (w,k)-minimizers (must be <=k) [default: 14]
  -k INT                      K-mer size for (w,k)-minimizers [default: 15]
  -T,--top Excludes: --input --bottom
                              Output the top path through each local PRG
  -B,--bottom Excludes: --input --top
                              Output the bottom path through each local PRG
  --flag Needs: --input       output success/fail rather than the node path
  -v                          Verbosity of logging. Repeat for increased verbosity
```

**Comment by mbhall88** [comment]
# get_vcf_ref

## Old

```
$ pandora get_vcf_ref -h
[2020-09-05 07:05:30.279251] [0x00007f0a41346f80] [debug]   Loading PRGs from file -h
terminate called after throwing an instance of 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >'
SIGABRT: abort
```
followed by a whole bunch of singularity crazy syscall error messages :grimacing: 

## New

```
$ pandora get_vcf_ref -h
Outputs a fasta suitable for use as the VCF reference using input sequences
Usage: ./pandora get_vcf_ref [OPTIONS] <PRG> [<QUERY>]

Positionals:
  <PRG> FILE [required]       PRG to index (in fasta format)
  <QUERY> FILE                Fast{a,q} file of sequences for some reason???

Options:
  -h,--help                   Print this help message and exit
  -z,--compress               Compress the output with gzip
  -v                          Verbosity of logging. Repeat for increased verbosity
```

**Comment by mbhall88** [comment]
# random_path

I would also like to suggest renaming this subcommand to just be `random`

## Old

```
$ pandora random_path -h
[2020-09-05 07:33:11.837477] [0x00007f34f049ff80] [debug]   Loading PRGs from file -h
terminate called after throwing an instance of 'std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >'
SIGABRT: abort
PC=0x47ce4b m=0 sigcode=0
```
followed by a whole bunch of error log

## New

```
$ pandora random_path -h
Outputs a fasta of random paths through the PRGs
Usage: ./pandora random_path [OPTIONS] <PRG>

Positionals:
  <PRG> FILE [required]       PRG to generate random paths from

Options:
  -h,--help                   Print this help message and exit
  -n INT                      Number of paths to output [default: 1]
  -z,--compress               Compress the output with gzip
  -v                          Verbosity of logging. Repeat for increased verbosity
```

**Comment by iqbal-lab** [comment]
Might it be clearer to rename "check_kmergraph" to - "seq_to_path", to  convert a sequence to a path in the graph?

**Comment by mbhall88** [comment]
> Might it be clearer to rename "check_kmergraph" to - "seq_to_path", to convert a sequence to a path in the graph?

Sure.

**Comment by mbhall88** [comment]
# merge_index

## Old

```
$ pandora merge_index -h
Usage: pandora merge_index <index1> <index2> ...
Options:
        --outfile                               Filename for merged index
```

## New

```
$ pandora merge_index -h
Allows multiple indicies to be merged (no compatibility check)
Usage: ./pandora merge_index [OPTIONS] <IDX>...

Positionals:
  <IDX> FILES [required]      Indicies to merge

Options:
  -h,--help                   Print this help message and exit
  -o,--outfile FILE           Filename for merged index [default: "merged_index.idx"]
  -v                          Verbosity of logging. Repeat for increased verbosity
```

**Comment by mbhall88** [comment]
Ok. That is all the architecture setup for the new CLI. All subcommands are now implemented. Reviewing this is a big job so feel free to do it in chunks. I think the best way is to leave comments in the code review rather than in the conversation here.  
I have left a whole bunch of `// todo:` comments scattered throughout in parts where we need a discussion.

**Comment by mbhall88** [comment]
I have updated some descriptions and shortened some options. 
I am about to start working on a PR to merge candidate regions [#228] so I would like to merge this CLI PR as soon as possible to avoid problems with the other PR comes in (i.e. conflicts with this CLI PR as it will be based off this work)

**Comment by leoisl** [comment]
I can try to review it tomorrow, but I am overloaded... might just be able to do it on Wednesday. I think it is better to merge, and then I review at some point where I don't have to rush the review. If there is any issue, I will then make a second CLI PR, but just from an overview of what you did, I think it should be all fine. I think my reviewing will be limited also to just look at the code, but we might not be able to capture possible bugs (if there are any), as most of the code you changed (files implementing main()) have no test coverage

**Comment by leoisl** [comment]
Hello again, sorry, will just be able review tomorrow :(

**Review by leoisl — APPROVED** [review]
All fine, just some very minor stuff! You can choose if implement them or not

File: src/get_vcf_ref_main.cpp [hunk 1]
```diff
@@ -1,71 +1,88 @@
...
+    auto gvr_subcmd = app.add_subcommand("get_vcf_ref", description);
+
+    gvr_subcmd->add_option("<PRG>", opt->prgfile, "PRG to index (in fasta format)")
+        ->required()
+        ->check(CLI::ExistingFile.description(""))
+        ->type_name("FILE");
+
+    gvr_subcmd
+        ->add_option(
+            "<QUERY>", opt->seqfile, "Fast{a,q} file of sequences for some reason???")
```
[the remark this answers is not in the document]
Comment (mbhall88) [reply src/get_vcf_ref_main.cpp:18]: Oh woops!

File: src/compare_main.cpp [hunk 2]
```diff
@@ -86,315 +185,76 @@ std::vector<std::pair<SampleIdText, SampleFpath>> load_read_index(
...
+    // todo: this all seems strange
+    if (opt.error_rate < 0.01) {
+        opt.illumina = true;
     }
-
-    assert(w <= k);
-    assert(not prgfile.empty());
-    if (illumina and e_rate > 0.1) {
-        e_rate = 0.001;
+    if (opt.error_rate > 0.05 and opt.illumina) {
```
[the remark this answers is not in the document]
Comment (mbhall88) [reply src/compare_main.cpp:221]: It was in the code before - it was further down the file. This whole section is weird! There seems to be so many contradictions. I just wanted to keep the current functionality the same. I suspect we will revisit this with #232

File: src/compare_main.cpp [hunk 3]
```diff
@@ -510,18 +369,20 @@ int pandora_compare(int argc, char* argv[])
...
+    for (uint32_t i = 0; i <= pangraph->nodes.size() / nb_vcfs_per_dir; ++i) {
+        fs::create_directories(vcfs_dir + "/" + int_to_string(i + 1));
+    }
 
     // create the dirs for the VCFs genotyped, if genotyping should be done
-    auto VCFsGenotypedDirs = outdir + "/VCFs_genotyped_" + genotype;
-    if (do_global_genotyping || do_local_genotyping) {
-        for (uint32_t i = 0; i <= pangraph->nodes.size() / nbOfVCFsPerDir; ++i)
-            fs::create_directories(VCFsGenotypedDirs + "/" + int_to_string(i + 1));
+    auto vcfs_genotyped_dirs = opt.outdir + "/VCFs_genotyped";
```
[the remark this answers is not in the document]
Comment (mbhall88) [reply src/compare_main.cpp:381]: Given we are saying local genotyping is for dev purposes it doesn't make sense to have global in the filename. Also, it makes the filename bloated with more info than is need IMO.

File: src/walk_main.cpp [hunk 4]
```diff
@@ -1,29 +1,42 @@
...
-            stderr, "Usage: pandora walk <in_prg.fa> [<seq.fa> | --top | --bottom]\n");
-        return 1;
-    }
+    auto opt = std::make_shared<WalkOptions>();
+    auto walk_subcmd = app.add_subcommand("walk",
+        "Outputs a path through the nodes in a PRG corresponding to the either an "
+        "input sequence (if it exists) or the top/bottom path");
+    walk_subcmd->add_option("<PRG>", opt->prgfile, "A PRG file (in fasta format)")
+        ->required()
+//        ->check(CLI::ExistingFile.description(""))
```
[the remark this answers is not in the document]
Comment (mbhall88) [reply src/walk_main.cpp:11]: Good pickup. Don't know what happened here...

File: src/compare_main.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (leoisl) [reply src/compare_main.cpp:220]: Ok, I think this is fine then! I guess it is a good practice to keep the same functionality in a refactoring, and documents issues to be fixed in next PRs!

File: src/compare_main.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (leoisl) [reply src/compare_main.cpp:380]: Ok!

**Comment by leoisl** [comment]
As a last task for this PR to be merged, should we try to run it in a real pipeline, e.g. in your TB pipeline? I think things are fine, but as there are no code coverage on most of these code changes, I'd like to ensure that at least our normal workflow is running correctly. This task could also be pushed to later, as these changes will eventually be tested in a normal workflow...

**Comment by mbhall88** [comment]
> As a last task for this PR to be merged, should we try to run it in a real pipeline, e.g. in your TB pipeline? I think things are fine, but as there are no code coverage on most of these code changes, I'd like to ensure that at least our normal workflow is running correctly. This task could also be pushed to later, as these changes will eventually be tested in a normal workflow...

Yeah, I've already been doing this.

**Pull request merged**